Penalty line tactic

hdemusg opened the pull request:
## Description
Updating the line tactic end points in the Penalty_Offense and Penalty_Defense tactics

## Associated Issue
**update line tactic start points for penalty situations according to rules** on ClickUp

## Steps to test and expected result
1. On penalty offense running, make sure 5 robots line up 7.5 metres away from the opponent's goal
2. On penalty defense running, make sure 4 robots line up 7.5 metres from our goal. 

7.5 metres is calculated per regulations (penalty line is 6 metres away, and all robots must be 1 metre behind the penalty line before the ball is kicked.

hdemusg committed:
Update penalty_offense.py

hdemusg committed:
Defense

hdemusg committed:
Added comment

hdemusg committed:
Update penalty_offense.py

hdemusg committed:
Update penalty_defense.py

hdemusg commented:
There's almost certainly issues with this, but I am currently unable to run the sim due to VM issues so I just did a quick fix and will improve the code once I've sorted the VM and sim issues in the next few days.

p-nayak11 commented on rj_gameplay/rj_gameplay/play/penalty_offense.py:
```diff
@@ -33,8 +33,8 @@ def tick(
-            start_pt = np.array([-2.0, 1.5])
-            end_pt = np.array([2.0, 1.5])
+            start_pt = np.array([-2.4, 1.5])
+            end_pt = np.array([2.4, 1.5])
```
Is there a particular reason why you widened the length of the line that the robots LineUp at?

hdemusg replied:
The only reason for that was to more evenly spread out the bots on the field but I guess it makes more sense for them to be bunched up tactically.

hdemusg replied:
I think I'll keep the robots more spread out in case the penalty is saved and moves towards the edges of the field.

p-nayak11 commented on rj_gameplay/rj_gameplay/play/penalty_defense.py:
```diff
@@ -29,8 +29,9 @@ def tick(
             num_liners = len(world_state.our_visible_robots) - 1
-            start_pt = np.array([-3.0, 0.5])
-            end_pt = np.array([-3.0, 5.5])
+            # 7.5 metres to our goal (esigned so that 5 robots can be at least 1 metre behind the penalty line before the penalty is taken)
```
```suggestion
            # 7.5 meters to our goal (designed so that 5 robots can be at least 1 meter behind the penalty line before the penalty is taken)
```

p-nayak11 commented on rj_gameplay/rj_gameplay/play/penalty_offense.py:
```diff
@@ -62,8 +62,9 @@ def tick(
             num_liners = len(world_state.our_visible_robots) - 1
-            start_pt = np.array([-2.0, 1.5])
-            end_pt = np.array([2.0, 1.5])
+            # 7.5 metres to opponent's goal (6 metres to penalty line, 1 metre behind line, then spare distance to be safe)
```
```suggestion
            # 7.5 meters to opponent's goal (6 meters to penalty line, 1 meter behind line, then spare distance to be safe)
```
I know it's another way to spell meter, but everything else in the codebase is spelled the American way lmao

p-nayak11 reviewed (changes requested):
Overall, the placement of the line is correct, and all of the robots line up appropriately according to the rules. I noticed that the goalie is not sticking to the goal lines during Penalty Defense, so that will be a card that I will add to clickup.

Some of the changes that I would like to see are mostly regarding documentation. Look at [this](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) page (specifically the Docstrings section) to document these files. Please include specific rules in the summary section and/or links to the rules regarding each situation. 

PS - Don't worry about adding the docstrings for the inherited methods since they are defined in the superclass. Only the classes need these docstrings.

hdemusg committed:
Update rj_gameplay/rj_gameplay/play/penalty_defense.py

Co-authored-by: p-nayak11 <44911049+p-nayak11@users.noreply.github.com>

hdemusg committed:
Update rj_gameplay/rj_gameplay/play/penalty_offense.py

Co-authored-by: p-nayak11 <44911049+p-nayak11@users.noreply.github.com>

hdemusg committed:
Docstring for penaltyDefense

hdemusg committed:
Docstring for PenaltyOffense

hdemusg committed:
Updated PenaltyDefense

I realized that the robots should be a bit more dispersed along the line to prepare to clear the penalty in the event of a save.

hdemusg committed:
Update penalty_offense.py

hdemusg commented:
Apologies for the delays. I have written the docstrings. Please let me know if I should make any edits to them.

p-nayak11 commented on rj_gameplay/rj_gameplay/play/penalty_defense.py:
```diff
@@ -10,13 +10,23 @@
 class PenaltyDefense(stp.play.Play):
+    """This is a play that represents the lineup procedure in the event that an opposing team has been awarded a penalty kick. 
+       Per rule 3 in Section 5.3.5 of the RoboCup SSL Playbook (https://robocup-ssl.github.io/ssl-rules/sslrules.html#_penalty_kick),
+       "Throughout the penalty kick procedure, all other robots have to be 1m behind the ball such that they do not interfere the penalty kick procedure".
+       As such, this play assigns one goalkeeper robot and lines up the remaining robots at least 1m behind the ball to prepare for the penalty. 
```
I would just refer to the rule than copy-paste it in.

p-nayak11 commented on rj_gameplay/rj_gameplay/play/penalty_offense.py:
```diff
@@ -12,13 +12,23 @@
-    """striker + line up rest of robots"""
+    """This is a play that represents the lineup procedure in the event that our team has been awarded a penalty kick. 
+       Per rule 3 in Section 5.3.5 of the RoboCup SSL Playbook (https://robocup-ssl.github.io/ssl-rules/sslrules.html#_penalty_kick),
+       "Throughout the penalty kick procedure, all other robots have to be 1m behind the ball such that they do not interfere the penalty kick procedure".
+       As such, this play lines up 5 offensive robots such that they are at least 1m behind the ball and assigns a striker to take the penalty.
+
```
Same as the comment above, don't need the whole rule just the reference.

p-nayak11 reviewed (changes requested):
I left some comments. There are some style errors that you must fix. Could you also wrap your comments to 80 characters according to the contributions guide? [Here](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) is an extension for vs code.

hdemusg committed:
Style changes and documentation change

hdemusg committed:
Style change and documentation
